diff --git a/src/libflake/flake/flake.cc b/src/libflake/flake/flake.cc
index 627dfe830..d7fef3307 100644
--- a/src/libflake/flake/flake.cc
+++ b/src/libflake/flake/flake.cc
@@ -79,7 +79,7 @@ static std::tuple<StorePath, FlakeRef, FlakeRef> fetchOrSubstituteTree(
 
 static void forceTrivialValue(EvalState & state, Value & value, const PosIdx pos)
 {
-    if (value.isThunk() && value.isTrivial())
+    if (value.isThunk())
         state.forceValue(value, pos);
 }
 
@@ -216,7 +216,7 @@ static Flake readFlake(
 
     // NOTE evalFile forces vInfo to be an attrset because mustBeTrivial is true.
     Value vInfo;
-    state.evalFile(flakePath, vInfo, true);
+    state.evalFile(flakePath, vInfo, false);
 
     Flake flake {
         .originalRef = originalRef,
